[PT2 Compiler] Add flag in dynamo disable wrapper to indicate reursive disable

malaybag opened the pull request:
Summary: After torch._dynamo.disable is applied, wrapped method does not have any flag to indicate whether it was disabled recursively or not. This flag is needed if to preserve dynamo disable methods in torch.export-ed model

Test Plan:
```
buck test mode/opt caffe2/test/dynamo:test_dynamo -- 'test_disable_recursive_flags'
````
https://www.internalfb.com/intern/testinfra/testrun/7599824674075603

Differential Revision: D84949143




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

angelayi reviewed (changes requested):
Can we hold off on landing this PR until we prototype out the entire "preserving disable in export"

malaybag commented:
> Can we hold off on landing this PR until we prototype out the entire "preserving disable in export"

Yes, that makes sense. Will ping when e2e implementation is ready.

williamwen42 commented on torch/_dynamo/decorators.py:
```diff
@@ -1002,3 +1003,10 @@ def error_on_graph_break(
+
+def is_dynamo_disable_recursive(method: Callable[[Any], Any]) -> Optional[bool]:
+    """
+    Check if a method is marked as `dynamo_disable` recursively.
```
Document what should be returned:
- True if disable(recursive=True)
- False if disable(recursive=False)
- None if method is not a disable decorator

williamwen42 reviewed (approved):
Just a small doc nit

malaybag committed:
[PT2 Compiler] Add flag in dynamo disable wrapper to indicate reursive disable (#165790)

Summary:

After torch._dynamo.disable is applied, wrapped method does not have any flag to indicate whether it was disabled recursively or not. This flag is needed if to preserve dynamo disable methods in torch.export-ed model

Test Plan:
```
buck test mode/opt caffe2/test/dynamo:test_dynamo -- 'test_disable_recursive_flags'
````
https://www.internalfb.com/intern/testinfra/testrun/7599824674075603

Reviewed By: williamwen42, Microve

Differential Revision: D84949143
